Ruby layer refactor and new shadowenv layer

syl20bnr opened the pull request:
Related issue: https://github.com/syl20bnr/spacemacs/issues/14653

* new macro `spacemacs|eval-for-enabled-ruby-mode`
* use "dispatch" pattern with local vars hook for backend, test-runner, version-manager
* use spacebind
* diminish minor modes lighter
* new layer: shadowenv
* add shadowenv layer to ruby mode, python mode and compilation buffer

Commit:
[ruby] Refactor layer

* new macro `spacemacs|eval-for-enabled-ruby-mode`
* use "dispatch" pattern with local vars hook for backend, test-runner, version-manager
* use spacebind

Commit:
[shadowenv] Create layer

Commit:
[shadowenv] Add shadowenv support for ruby and compilation modes

Commit:
[shadowenv] Add shadowenv support for python mode

Commit:
Update change log

lebensterben commented on layers/+tools/shadowenv/README.org:
```diff
@@ -0,0 +1,31 @@
+
+#+TAGS: layer|tool
+
+[[file:img/shadowenv.png]]
```
I suggest to write copyright notice for all image files from now on.

So add a footnote `[[file:img/shadowenv.png]] [fn:1]`
And `[fn:1] This logo image consists only of simple geometric shapes or text. It does not meet the threshold of originality needed for copyright protection, and is therefore in the public domain.`

syl20bnr replied:
:+1: done.

lebensterben commented on layers/+tools/shadowenv/packages.el:
```diff
@@ -0,0 +1,40 @@
+
+(defun shadowenv/init-shadowenv ()
+  (use-package shadowenv
+    :defer t
```
optionally we can use
`:spacediminish " Ⓥ" " V"`

It takes 0 to 3 arguments, same as those of `spacemacs|diminish`.
When the first argument is omitted, it's inferred from the package name, which is `(intern (format "%s-mode" PKG-NAME))`.

syl20bnr replied:
Thanks for mentioning `:spacediminish`, I was not aware of it.

Done.

lebensterben commented on layers/+tools/shadowenv/packages.el:
```diff
@@ -0,0 +1,40 @@
+      (spacemacs|diminish shadowenv-mode " Ⓥ" " V")
+      (eval `(spacemacs|spacebind
+              :major
+              (,mode
```
A think this is a common case. `spacemacs|spacebind` should accept either a symbol or a list of symbols as for `MODE`.

JAremko replied:
@lebensterben  Already planed :smile:   We have a bunch of layers doing this kind of templating.

syl20bnr replied:
I did not think about static analysis. I guess we should drop the idea of static analysis because it will be like mission impossible to catch all the calls to `spacebind` correctly. For example here, even if I drop the macro to replace with a `dolist` (which would an inferior solution) the static analysis will have a hard time to detect each iteration of the `spacebind` call.
A better place for static analysis could be the `README.org` files once the key bindings have been rendered into them.

syl20bnr replied:
Makes me think that the macro here will prevent from correctly rendering in the `README.org`. 
You are right, passing an explicit list of modes is the way to do.

syl20bnr replied:
OK so we can keep it like it for now. When we support a list of modes we will migrate the spacebind calls out of the macro.

lebensterben commented on layers/+lang/ruby/funcs.el:
```diff
@@ -160,9 +225,5 @@ Called interactively it prompts for a directory."
-(defun spacemacs/ruby-fmt-before-save-hook ()
-  (add-hook 'before-save-hook 'spacemacs/ruby-format t t))
+(defun spacemacs//ruby-add-prettier-js-before-save-hook ()
+  (add-hook 'before-save-hook 'prettier-js t t))
```
the third argument to `add-hook` is `DEPTH`.
It defaults to 0, and any non-nil symbol implies 90.
So this hook is likely to be run after other hooks (-100 is first and 100 is last).

Is this really the behaviour we want?

syl20bnr replied:
I guess so. The addition of `DEPTH` is a rather new one and it is meant to be backward compatible. 
If there was `t` to append, then `DEPTH` should keep the same behavior.

lebensterben commented on layers/+lang/ruby/funcs.el:
```diff
@@ -22,11 +22,55 @@
+      (progn
+        (cl-nsubstitute-if new 'spacemacs//ruby-substitute-predicate body)
+        nil)
+    (eq body old)))
```
`old` is dynamically scoped. It's better to make it a function argument.

syl20bnr replied:
I thought about it but I could not do that, the predicate for `cl-substitute` is a function with one argument.
What could be done to improve upon this ? Sort of a curried function ? Or maybe declare variables explicitly ? Another macro ?

lebensterben replied:
Solution one

well. let's make it to take pair, (OLD . BODY).
it's a function taking ONE argument.
`(cl-nsubstitute-if new 'spacemacs//ruby-substitute-predicate (cons old body)`

---

Make it take two arguments, the first is `old`
Solution two `(cl-nsubstitute-if new (apply-partially old) body)`.

lebensterben commented on layers/+lang/ruby/funcs.el:
```diff
@@ -22,11 +22,55 @@
+      (cl-nsubstitute-if new 'spacemacs//ruby-substitute-predicate body))
+    (let ((new 'ruby-mode-local-vars-hook)
+          (old 'local-vars-hook))
+      (cl-nsubstitute-if new 'spacemacs//ruby-substitute-predicate body))))
```
This part can be simplified.
First we let-bound a list of modes based on whether `ruby-enable-enh-ruby-mode` is t.
```elisp
(let* ((rubymode (if ruby-enable-enh-ruby-mode
                     "enh-ruby"
                  "ruby")
       (pairs (cl-loop for old in '("mode" "hook" "local-vars-hook")
                       for new = (s-join "-" (list rubymode old))
                       collect (cons old new)))
 ..)
```
Then we can simply
```elisp
  (dolist (pair pairs)
     (cl-nsubstitute-if (cdr pair) 'spacemacs//ruby-substitute-predicate (car pair)))
```

lebensterben replied:
The real question is, how efficient is `cl-nsubstitute-if`.

syl20bnr replied:
Good question, I would assume it is the fastest as it is the destructive version `cl-substitute-if` but this is a wild guess 😅 
Going to update with your great implementation for now.

syl20bnr replied:
Done

Commit:
[conventions] Add mode-line ligther conventions for toggles

The reason why I moved to parenthesized letters is because parenthesized
symbols exist for both small and capital letters.
This is not the case with negative circled and negative squared letters.

Commit:
Apply new mode-line lighter convention for major mode toggles

lebensterben commented on layers/+lang/ruby/funcs.el:
```diff
@@ -97,39 +150,49 @@
 ;; rspec
+(defun spacemacs//ruby-setup-rspec ()
+  "Setup rspec test runner."
+  (when (eq ' 'rvm)
```
typo?

syl20bnr replied:
Fixed.

syl20bnr commented:
I updated the conventions for toggles lighters in the mode line, see commit message https://github.com/syl20bnr/spacemacs/commit/c8ae548c9
These conventions were missing from the convention doc, this commit adds them.

Kazark commented on layers/+tools/shadowenv/README.org:
```diff
@@ -0,0 +1,31 @@
+This layer adds support for [[https://shopify.github.io/shadowenv/][shadowenv]]
+
+** Features:
+- Automatically activatefouhnd shadowenv each supported buffer locally
```
typo: `activatefouhnd`

syl20bnr replied:
Fixed.

Commit:
[ruby] Refactor macro spacemacs|eval-for-enabled-ruby-mode

Commit:
[shadowenv] Add footnote about image copyright

Commit:
[core] Fix missing which-key description major mode leader key

Added support for `,` leader key in which-key descriptions.

lebensterben commented:
Also it'd be nice if you can change
`(add-hook 'foo 'bar)` to `(add-hook 'foo #'bar)` where bar is a symbol of function.

This is not strictly enforced now but is easier for byte compiler to understand such symbol is a function.

Commit:
[core] Fix missing which-key description for minor mode

Commit:
[shadowenv] Bind key bindings to minor mode

This allows to activate the key bindings only when the
shadowenv minor mode is enabled.

Commit:
[shell] Add support for shadowenv

Commit:
[Conventions] Add SPC V prefix for Version/Environment

Commit:
[shadowenv] Use global shadowenv mode instead of local

Move the key bindings to global keymap and follow the new SPC V
key binding convention.

Commit:
[defaults] Add SPC V e prefix for Emacs environments

Commit:
[finance] Better way to enable flycheck gobally when visiting ledger files

smile13241324 reviewed (changes requested):
Just reviewed your PR. I have found some smaller typos which should be changed before it can be merged.

syl20bnr commented:
Do not merge this yet, I'm stil working on it.

Commit:
[core] New leader key! Project leader key.

This new leader is a whole new scope for modes and key bindings in
Spacemacs.

The current scope are `global` and `major-mode`. The `major-mode`
scope has a dedicated leader `SPC m` or `,` as shortcut.

This commits adds a new scope: `project`. It hosts key bindings for
minor modes enabled for a project (in the projectile sense). This
was a missing piece in Spacemacs to correctly handle minor modes
like Ruby On Rails framework key bindings. These key bindings need
to be available in the whole project no matter the current major
mode.

The new scope has a new dedicated leader key: `SPC M` and `;` as
shortcut. `;` was chosen because it is the symmetrical of `,` in
VIM thus it makes sense to use it.

* new dotspacemacs variables:
    dotspacemacs-project-minor-mode-leader-key
    dotspacemacs-project-minor-mode-emacs-leader-key
* new functions in `core-keybindings.el`:
    spacemacs/declare-prefix-for-project-minor-mode
    spacemacs//init-project-minor-mode-leader-mode-map
    spacemacs/set-leader-keys-for-project-minor-mode
* new keyword in spacebind: :project-minor
* new command prefix: "M" for "Project commands"

Commit:
[core] New local variable concept: project type

* New file: core-local-variables.el
* New interactive function in Helm layer: spacemacs/helm-add-project-type

This new file expands the hack local variables capabilities of
Spacemacs with a new concept: Project Types

Each layer can add a project type with the function
`spacemacs/add-available-project-type`

With the function `` a user can add a type to the current project.

When a file is opened all the functions bound to the corresponding
project type using the function `spacemacs/add-local-var-hook` will
be executed (they are essentially setup functions to enable minor
modes conditionally).

The user can easily add a type for the current project with `SPC p s`
to set a type. It effectively adds the selected type to the `.dir-locals.el`
file at the root of the project.

Commit:
[python] [WIP] Python layer refactor

Commit:
[evil] Enable evil-snipe by default and add documentation

Since the project leader is on `;` we now enable evil-snipe by
default to work-around some of the functionality loss.

* delete the layer `evil-snipe`
* move the init function of evil-snipe to `spacemacs-evil` layer
* move additional configuration to `git` and `ranger` layers
* add documentation about evil-snipe in DOCUMENTATION.org

Commit:
[ruby] Add project type "ruby"

Commit:
[ruby-on-rails] Add project type "ruby-on-rails"

JAremko commented on layers/+spacemacs/spacemacs-defaults/keybindings.el:
```diff
@@ -733,6 +734,15 @@ respond to this toggle."
+(spacemacs|spacebind
+ "Version/Environment"
+ :global
+ (("V" "Version/Environment"
```
@syl20bnr  Extra set of `()`


@lebensterben  I think we should detect it and report as en error since it is easy to introduce when refactoring.

Commit:
[doc] Update README for installation of Railwaycat's Emacs

Commit:
Update documentation for project type

Commit:
[lsp] Add SPC m g D to open definition in other window

syl20bnr committed:
Run local var major hook after the project type one

syl20bnr committed:
Add project type support in vue layer

syl20bnr committed:
Add project type support in javascript layer

syl20bnr committed:
Add project type support in typescript layer

syl20bnr committed:
Fix lsp server conflict when multiple languages in the same project

syl20bnr committed:
[compleseus] Tweak compleseus user interface

* Add C-z support to narrow down embark actions with completing-read (similar
to helm)
* Add helm-ag workflow to replace occurrences in multiple files easily using
C-c C-e
* Update the documentation to reflect the changes

syl20bnr committed:
Add missing guards in git gutter package configurations

syl20bnr committed:
[compleseus] Update README

syl20bnr committed:
[compleseus] Replace consult-multi-occur wuth consult-line-multi

Fixes #16146

smile13241324 commented:
Given that this is now older than 3 years I would like to go ahead and close it for now.
